File: python/step3.py
import requests

'''
Step 3: Needle in a haystack
------------------------------
We’re going to send you a dictionary with two values and keys.
The first value, needle, is a string. The second value, haystack, is an array of strings.
You’re going to tell the API where the needle is in the array.


This is probably the first time I've worked with dictionaries and their key/val pairs!
Once I found out how to store the val's in their respective variables, the rest was a piece of cake!
I also found out that Python has its own "indexOf()" function.
'''

def findNeedle(needle, haystack):
	for key in haystack:
		if key == needle:
			return haystack.index(needle)

token = {'token':'4f36c1b5089b07d3416239291874b956'}
r = requests.post('http://challenge.code2040.org/api/haystack', json=token)

info = r.json()
needle = info['needle']
haystack = info['haystack']

# print(findNeedle(needle, haystack))

result = {'token':'4f36c1b5089b07d3416239291874b956', 'needle':findNeedle(needle, haystack)}
s = requests.post('http://challenge.code2040.org/api/haystack/validate', json=result)
print(s.status_code)

File: go/step3.go
package main

/*
Step 3: Needle in a haystack
------------------------------
We’re going to send you a dictionary with two values and keys.
The first value, needle, is a string. The second value, haystack, is an array of strings.
You’re going to tell the API where the needle is in the array.

Implementing this in Go was fun, but alot of work. I'm still learning the in's and out's of the language
*/

import (
	"fmt"
	"github.com/jochasinga/requests"
	"io/ioutil"
	"encoding/json"
)

func findNeedle(needle string, haystack []string) int {
	var temp int
	for key, val := range haystack {
		if needle == val {
			temp = key
		}
	}
	return temp
}

type Params struct {
	Token string `json:"token,omitempty"`
	Index int `json:"index,omitempty"`
}

type NeedleHayStack struct {
	Needle   string `json:"needle"`
	Haystack []string `json:"haystack"`
}

func main() {
	token := &Params{
		Token: "4f36c1b5089b07d3416239291874b956",
	}

	res, err := requests.PostJSON("http://challenge.code2040.org/api/haystack", token)
	if err != nil {
		panic(err)
	}

	// store as []byte
	page, err := ioutil.ReadAll(res.Body)
	res.Body.Close()
	if err != nil {
		panic(err)
	}

	var test NeedleHayStack
	json.Unmarshal(page, &test)

	// fmt.Println(test)
	need := test.Needle
	hay := test.Haystack

	result := &Params {
		Token: "4f36c1b5089b07d3416239291874b956",
		Index: findNeedle(need, hay),
	}

	results, err := requests.PostJSON("http://challenge.code2040.org/api/haystack/validate", result)
	if err != nil {
		panic(err)
	}

	fmt.Println("index of needle: ", findNeedle(need, hay))
	fmt.Println(results.StatusCode) // post returns 400....

}
